File: backEnd/src/main/java/tn/esprit/backend/entity/Property.java
package tn.esprit.backend.entity;

import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.FieldDefaults;

import java.io.Serializable;
import java.util.Date;
import java.util.List;

@Entity
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Builder
@ToString
@FieldDefaults(level = AccessLevel.PRIVATE)
public class Property implements Serializable {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Setter
    private Long propertyId;

    private String title;

    private String description;

    private Double price;

    private String location;

    @Enumerated(EnumType.STRING)
    private PropertyType type;

    @Enumerated(EnumType.STRING)
    private PropertyStatus status;

    @Temporal(TemporalType.DATE)
    private Date postedDate;

    @Lob
    @Column(name = "image", columnDefinition = "MEDIUMBLOB")
    private byte[] image;

    @Transient
    private Double averageRating;

    // @ManyToOne
    // @ToString.Exclude
    // @JsonIgnore
    // User user;

    @OneToMany(mappedBy = "property", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<Feedback> feedbacks;
}

File: frontEnd/src/app/models/Property.ts
import { Feedback } from "./Feedback";
export class Property {
    propertyId!:number;
    title!:string;
    description!:string;
    price!:number;
    location!:string;
    type!:PropertyType;
    status!:PropertyStatus;
    postedDate:Date;
    image!:string;
    averageRating!:number;
    feedbacks!: Feedback[];
    constructor(
        propertyId: number,
        title: string,
        description: string,
        price: number,
        location: string,
        type: PropertyType,
        status: PropertyStatus,
        postedDate: Date,
        image: string,
        feedbacks?: Feedback[],
        averageRating?: number
      ) {
        this.propertyId = propertyId;
        this.title = title;
        this.description = description;
        this.price = price;
        this.location = location;
        this.type = type;
        this.status = status;
        this.postedDate = postedDate;
        this.image = image;
        this.feedbacks = feedbacks || [];
        this.averageRating = averageRating || 0;
      }
}


export enum PropertyType {
  Sale = 'Sale',
  Rent = 'Rent'
}

export enum PropertyStatus {
  Available = 'Available',
  Sold = 'Sold'
}